Fix: Enforce consistent Maintenance Mode Strategy behavior

m-ildefons opened the pull request:
# Problem:

As described in harvester/harvester#6835, there are several problems to be considered here:

1. Ensuring that the maintenance-mode strategy is specified
2. Ensuring that only a valid maintenance-mode strategy can be specified
3. Ensuring that the specified maintenance-mode strategy label is propagated all the way to all resources where it's expected
4. Ensuring that default behavior is applied in all cases where a maintenance-mode strategy either isn't specified or isn't valid

# Solution:

The solution to the problem consists of a modification to the node drain helper and modifications both the mutating admission webhook and the validating admission webhook for the VirtualMachine resource

### The Node Drain Helper

The node drain helper was only marking Pods of VMs which had the maintenance-mode strategy label set to `Migrate` for deletion (i.e. migration off of the node).
This behavior would omit all Pods which didn't have this label set at all, or the ones which had it set to an invalid value. Similarly, Pods of VMs which didn't have one of the three other maintenance-mode strategies set would not be shut down. Thereby any VM without a maintenance-mode strategy, or with an invalid maintenance-mode strategy would be left running, preventing the node from entering the maintenance mode successfully. As a result the node would remain stuck in "Cordoned" state, waiting for an operator to intervene and deal with the VMs.
The new behavior of the drain helper is to migrate any VMs off of the node, if the associated Pod either has no maintenance-mode strategy label, the maintenance-mode strategy label is invalid, or the maintenance-mode strategy is specified as `Migrate`. This is the documented default behavior for VMs.
All other VMs will have a different maintenance-mode strategy specified and they will thus be shut down (and later restarted according to their maintenance-mode strategy).

### Admission Webhook

The admission webhook for the VirtualMachine resource plays a crucial role in making sure that VirtualMachiness have a sane maintenance-mode strategy specified.
First, the mutating webhook ensures that the maintenance-mode strategy label is set on the VirtualMachine resource. If it isn't, it will be set to `Migrate`, which is the default.
Then the mutating webhook ensures that the maintenance-mode strategy label that is set on the VirtualMachine resource is propagated into the `.spec.templates.metadata.labels`, which is to ensure that the label is propagated throughout both the VirtualMachineInstance as well as the Pod belonging to that VirtualMachine.
The mutating webhook will not modify exising maintenance-mode strategy labels on existing VirtualMachine resources. While the node drain helper will apply the default strategy to these virtual machines, this is to ensure that the user will not be faced with inexplicably changing maintenance-mode strategy labels, if they have an invalid value in their cluster.
The webhook will however log an appropriate message, allowing the user to identify VirtualMachine resources affected by invalid values in the maintenance-mode strategy label and correct it.

The validating webhook takes the role of a safe guard, ensuring that no VirtualMachine can be created or updated with an ill-specified maintenance-mode strategy.
It will reject the creation of VirtualMachine resources without a maintenance-mode strategy label or with an invalid value in the maintenance-mode strategy label.
This ensures that the label can not accidentally be filled with bogus data and that the user is informed.

# Related Issue:

- harvester/harvester#6835

# Test plan:

## Scenario 1

Ensure that new VirtualMachines have a valid maintenance-mode strategy. If it's not given, the default should be applied.

### Test 1

No maintenance-mode strategy given.

#### Setup:

Create a new VirtualMachine and specify no maintenance-mode strategy:
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/creator: harvester
    harvesterhci.io/os: linux
  name: test
[...]
```

#### Expected Result

The default strategy `Migrate` should have been added by the mutating admission webhook:
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/maintain-mode-strategy: Migrate
    [...]
  name: test
[...]
```

### Test 2

Valid maintenance-mode strategy given.

#### Setup:

Create a new VirtualMachine and specify a valid maintenance-mode strategy:
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/maintain-mode-strategy: Shutdown
    [...]
  name: test
[...]
```

#### Expected Result

The previously specified maintenance-mode strategy should be persisted in the VirtualMachine resource's metadata:
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/maintain-mode-strategy: Shutdown
    [...]
  name: test
[...]
```

## Scenario 2

If an invalid maintenance-mode strategy is specified, the VirtualMachine resource should be rejected.

### Test 1

Invalid maintenance-mode strategy given.

#### Setup:

Create a new VirtualMachine and specify an invalid maintenance-mode strategy:
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/maintain-mode-strategy: foobar
    [...]
  name: test
[...]
```

#### Expected Result

The creation should be rejected with a sensible error message

```
The request is invalid: metadata.labels[harvesterhci.io/maintain-mode-strategy]: invalid maintenance mode strategy: foobar
```

## Scenario 3

The specified maintenance-mode strategy should be propagated to all relevant resources.

### Test 1

The maintenance-mode strategy is propagated to the VirtualMachineInstance and the corresponding Pod, when the VirtualMachine is created

#### Setup

Create a new VirtualMachine and specify a valid maintenance-mode strategy:
```
type: kubevirt.io.virtualmachine
metadata:
  namespace: default
  labels:
    harvesterhci.io/maintain-mode-strategy: Shutdown
    [...]
  name: test
[...]
```

#### Expected Result

The VirtualMachine resource, the VirtualMachineInstance resource and the belonging Pod resource all have the specified maintenance-mode strategy label.

```
│ ~ │► kubectl get vm,vmi,pod -l harvesterhci.io/maintain-mode-strategy=Shutdown
NAME                              AGE   STATUS    READY
virtualmachine.kubevirt.io/test   10m   Running   True

NAME                                      AGE   PHASE     IP           NODENAME           READY
virtualmachineinstance.kubevirt.io/test   10m   Running   10.52.1.65   harvester-node-1   True

NAME                           READY   STATUS    RESTARTS   AGE
pod/virt-launcher-test-n5tf2   2/2     Running   0          10m
```

### Test 2

The maintenance-mode strategy is updated on the VirtualMachineInstance and on the corresponding Pod, when the maintenance-mode strategy is updated on the VirtualMachine resource.

//TODO

## Scenario 4

In case a VirtualMachine resource, VirtualMachineInstance resource, or a corresponding Pod exists and either has no maintenance-mode strategy label, or the maintenance-mode strategy label has an invalid value, the default behavior (`Migrate`) should be applied.

// TODO

m-ildefons committed:
Node Drain: Fix Pod deletion filter

The Pod deletion filter identifies the Pods belonging to VMs, which need
to be deleted during a node drain by looking at their maintenance-mode
strategy label. The default value for this label is `Migrate`, which
indicates that the Pod should be deleted during the node drain.
Pods with an invalid label, i.e. where the value of the maintenance-mode
strategy label is not one of:
  - Migrate
  - ShutdownAndRestartAfterEnable
  - ShutdownAndRestartAfterDisable
  - Shutdown
are now also being treated with the default behavior.
This fixes the problem that if a VM contains an invalied value in this
label, nodes can become stuck in `Cordoned` state when transitioning to
maintenance mode, as the node drain controller won't shut down the VM,
but it's also not migrated away from the node. Therefore the VM keeps
running, preventing the node from completely transitioning into
maintenance mode.

related-to: harvester/harvester#6835

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
VM Webhooks: Check Maintenance Mode Strategy

Add checks for maintenance-mode strategy to VM webhooks.

The Admission webhooks for the VirtualMachine resource needs to make
sure that the maintenance-mode strategy for the VM is set to a sane
value.
To do this, there are two checks:

The first one is in the mutating webhook, which is executed first, and
it just makes sure that the label, which defines the maintenance-mode
strategy, is set. If the label is not set, the mutating webhook will set
it with the default value of `Migrate`. If the label is set, the
mutating webhook will not modify it, even if it has an invalid value.
This ensures that the maintenance-mode strategy is never set
unintentionally to a wrong value.
The second check will ensure that in this case the request is rejected
with an error message, so the user can correct the value of the
maintenance-mode strategy label.
The mutating webhook will also ensure that the maintenance-mode strategy
label is copied from the `.metadata.labels` to
`.spec.template.metadata.labels`. This is necessary to ensure that the
Pod in which the virtual machine will run will be labeled correctly.

The second check happens in the validating webhook. This check ensures
that the maintenance-mode strategy label is set to a valid
maintenance-mode strategy, i.e. one of the values:
  - Migrate
  - ShutdownAndRestartAfterEnable
  - ShutdownAndRestartAfterDisable
  - Shutdown
This webhook will deny a CREATE or UPDATE, if the new VirtualMachine
resource does not contain the maintenance-mode strategy label at all, or
if it contains an invalid value.
The only exception is an UPDATE to a VirtualMachine resource that
already contains an invalid value in the maintenance-mode strategy label
and where the value does not change. In this case, the webhook will
accept the request. This is crucial, in case the controller needs to
deal with an existing VM that has an invalid value in this label (e.g.
on a cluster that has been upgraded from an old version, before this
label was checked by the admission webhook). In this case, the
controller still needs to be able to perform UPDATE operations on the
resource, to operate the VM.

Together, these two checks ensure that no VirtualMachine resource can be
created with an invalid maintenance-mode strategy, or with no
maintenance-mode strategy at all.
They also make sure that the maintenance-mode strategy can not be
removed or changed to an invalid value for existing VirtualMachine
resources.

related-to: harvester/harvester#6835

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
VMI Controller: Sync VMI labels to Pod

Add new VMI controller, which syncs all labels, which have the
`harvesterhci.io` prefix Pod(s) corresponding to the VMI.
This new controller ensures that Harvester's label namespace is
propagated all the way down to the Pod of a VirtualMachine.
Changes to the labels of a VirtualMachine or VirtualMachineInstance do
not justify re-creating the Pod, since labels are metadata. But labels
like the maintenance-mode strategy label have influence on the behavior
of VMs, and they may also be needed for features like load balancers to
work properly.
This controller syncronizes the labels from the VirtualMachineInstance
to the Pod to make sure that any change on the upstream objects is
reflected on the Pod.
For the maintenance-mode strategy label specifically, this means that
the maintenance-mode strategy of a VM can be updated without needing to
restart the VM.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

w13915984028 reviewed (commented):
Please check those questions, thanks.

m-ildefons committed:
VM Validataion Webhook: Simplify Flow

Simplify control flow for the maintenance-mode strategy label validation
in the VirtualMachine validating admission webhook.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
VMI Controller: Improvements to Label Syncing

Improve SyncHarvesterVMILabelsToPod control loop:

- Make use of the `maps` module from the standard library for cloning,
  comparing, copying. Also use `slices` module from the standard library
  for similar operations on arrays.
- Slight performance improvement in for loops over maps by avoiding
  double lookups of values
- Use differen Pod indexer to reduce the number of Pods needed to look
  through to find the Pods belonging to the VMI
- Check that labels really need updating before committing to expensive
  operations like DeepCopy and Update on the Pods
- Remove ad-hoc allocation and dynamic growing of array

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>
